src/Controller.php

```diff
@@ -49,85 +87,118 @@ public function setUp(\atk4\data\Model $m)
             return $a;
         });
 
+        // adds custom log methods in model
+        // log() method can clash with some debug logger, so we use two methods just in case
         if (!$m->hasMethod('log')) {
             $m->addMethod('log', [$this, 'customLog']);
         }
+        if (!$m->hasMethod('auditLog')) {
+            $m->addMethod('auditLog', [$this, 'customLog']);
+        }
 
+        // adds link to audit controller in model properties
         $m->audit_log_controller = $this;
     }
 
-    public function init()
+    /**
+     * Push change into audit log (and audit log stack).
+     *
+     * @param \atk4\data\Model $m
+     * @param string           $action
+     *
+     * @return model\AuditLog
+     */
+    public function push(\atk4\data\Model $m, $action)
     {
-        $this->_init();
+        // add audit model
+        $a = $m->ref('AuditLog');
 
-        $this->setUp($this->owner);
-    }
+        // set audit record values
+        $a['ts'] = new \DateTime();
 
-    public function push(\atk4\data\Model $m, $action)
-    {
-        if (isset($m->audit_model)) {
-            $a = clone $m->audit_model;
-        } else {
-            $a = clone $this->audit_model;
-        }
-        $m->persistence->add($a);
+        // sometimes we already have conditions set on model, but there are strange cases,
+        // when they are not. That's why we needed following 2 lines :(
+        // BUT hopefully don't need them anymore - let's see.
+        //$a['model'] = get_class($m);
+        //$a['model_id'] = $m->id;
 
         if ($this->custom_action) {
             $action = $this->custom_action;
             $this->custom_action = null;
         }
-
-        $a['ts'] = new \DateTime();
-        $a['model'] = get_class($m);
-        $a['model_id'] = $m->id;
         $a['action'] = $action;
 
         if ($this->custom_fields) {
             $a->set($this->custom_fields);
             $this->custom_fields = [];
         }
 
-        if (!$this->first_audit_log) {
-            $this->first_audit_log = $a;
-        }
-
         if ($this->audit_log_stack) {
+            // link to previous audit record
             $a['initiator_audit_log_id'] = $this->audit_log_stack[0]->id;
         }
 
         // save the initial action
         $a->save();
 
+        // memorize start time
         $a->start_mt = (float)microtime();
-        $m->audit_log = $a;
 
+        //Imants: deprecated - use $m->audit_log_controller->audit_log_stack[0] instead
+        // or $m->audit_log_controller->custom_action and custom_fields properties in your beforeSave hook
+        //$m->audit_log = $a;
```

Comment:
why not keep a reference to `audit_log` ?

what if there are multiple audit logs?

Reply:
I intentionally didn't want developer to mess directly with audit data model. He should use audit controller for that and audit controller should take care of audit data model.
I think that's better work-flow.